File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let l = 0
    let r = nums.length -1

    while(l<=r){
        const mid = Math.floor(l + (r - l) / 2)
        if(nums[mid] === target){
            return mid
        }

        // mid is in left half // right is +++/---
        if(nums[l] <= nums[mid]){ // use l pointers for both if statements
            if(target < nums[l] || target > nums[mid]){ // search right
                l = mid + 1
            } else {
                r = mid - 1
            }
        } else { // mid is in right half // left is +++/---
            if(target > nums[r] || target < nums[mid]){ // search left
                r = mid - 1
            } else {
                l = mid + 1
            }
        }
    }
    return -1
    
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        def helper(start, end):
            if start > end:
                return -1
            mid = (end+start)//2
            if nums[mid] == target:
                return mid
            
            if nums[start] <= nums[mid]:
                if nums[start] <= target < nums[mid]:
                    return helper(start, mid-1)
                else:
                    return helper(mid+1, end)
            else:
                if nums[mid] < target <= nums[end]:
                    return helper(mid+1, end)
                else:
                    return helper(start, mid-1)
        
        return helper(0, len(nums)-1)
        